src/token_bucket.rs

```diff
@@ -71,6 +57,53 @@ pub struct TokenBucket {
     client: Client,
 }
 
+async fn schedule_and_sleep(receiver: Receiver<ThreadState>) -> SLResult<()> {
+    // Receive class state
+    let ts = receiver.recv()?;
+
+    // Connect to redis
+    let mut connection = ts.client.get_async_connection().await?;
+
+    // Retrieve slot
+    let slot: u64 = get_script("src/scripts/schedule.lua")?
+        .key(&ts.name)
+        .arg(ts.capacity) // capacity
+        .arg(ts.frequency * 1000.0) // refill rate in ms
+        .arg(ts.amount) // refill amount
+        .invoke_async(&mut connection)
+        .await?;
```

Comment:
The return type of the redis call is `RedisResult` which means `Result<T, RedisError>`. 

We could unpack the result using 

```rust
match some_call.await {
    Ok(r) => r,
    Err(e) => {
        return PyValueError::new(...)
    }
}
```

(pyo3 requires that we return a PyErr from our "python" methods in rust)

but if you use the `?` operator, it will just surface the error (like re-raising in Python) to the next level up. That wouldn't work on it's own, since a `RedisError` from the redis crate isn't a `PyErr`, but by setting the return type to `Result<T, SLError>` where `SLError` is defined [here](https://github.com/sondrelg/self-limiters/blob/main/src/_errors.rs#L21), rust will try to convert the `RedisError` to an `SLError` for us 🤯 All we then have to do to make that conversion work, is tell the compiler how to do it, like [this](https://github.com/sondrelg/self-limiters/blob/main/src/_errors.rs#L41).

So tl;dr the `?` here makes it so that when the redis crate errors, a `RedisError` python exception is automatically raised

Reply:
Note that the crate has a `RedisError` and we've also created a python exception called `RedisError`, so in `_errors.rs` the crate error is imported as `RedisLibError`

Reply:
That’s so cool, thanks for explaining ❤️